### PR TITLE
feat(kubernetes): update kubernetes plugin to support matchLabels

### LINKED ISSUE #2924: k8s-plugin: Allow fetching of resources with arbitrary labels
From https://github.com/spotify/backstage/issues/2857#issuecomment-709753137

## Feature Suggestion

It should be possible to provide other labels, keys and values to look up resources by with a [label selector](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) like so `key=value`.

## Possible Implementation

I can see 2 ways of doing this:

### Entity driven

We could add another annotation which provides labels selectors to use for an entity such as:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: dice-roller
  description: It rolls dice
  tags:
    - go
  annotations:
    "backstage.io/kubernetes-id": "dice-roller"
spec:
  selector:
    matchLabels:
      name: dice-roller
      version: canary
  type: service
  lifecycle: production
  owner: guest

```

We would propagate this annotation from the frontend to the backend for each K8S fetch and construct a label selector like so:

`name=dice-roller&version=canary`

### Config Driven

Another option would be to provide config to the backend to tell it which field in the entity to use, for example:

For the Entity:

```yaml
apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: dice-roller
  description: It rolls dice
  tags:
    - go
spec:
  type: service
  lifecycle: production
  owner: guest
```

and the backend config: 

```yaml
...
kubernetes:
  labelSelectors:
    - key: "name"
      valueFrom: "entity.metadata.name"
    - key: "version"
      value: "canary"
...
```

We would construct the label `name=dice-roller&version=canary`

## Context

Currently we only allow users to to look up kubernetes resources if:

* The entity has an annotation like so `"backstage.io/kubernetes-id": entity_name`
* The Kubernetes resources to fetch have labels which match this value 

We then [fetch the the resources with a label selector](https://github.com/spotify/backstage/blob/master/plugins/kubernetes-backend/src/service/KubernetesFetcher.ts#L232).

This is quite limiting as users with then have to re-label their resources.

### PR DESCRIPTION
**Pull request by moustafab**
## Hey, I just made a Pull Request!

closes #2924 

I did a bit of refactoring to make things clean and moved some of the concern around building the labelSelector query to the front end, hopefully making it a bit clearer what exactly the request made to the k8s api is with the labelSelector (like other clients). There is room for future improvements to support other parts of labelSelector object, but this is a good start.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by freben** [comment]
Hey! Still some compilation errors happening.

**Comment by benjdlambert** [comment]
Hey @mclarke47 wanna jump in on a review here too?

**Comment by mclarke47** [comment]
LGTM so far!

**Comment by moustafab** [comment]
I'm not sure how to handle the entity modification, any tips or docs you can point me to?

File: plugins/kubernetes/src/api/types.ts
Comment (Rugvip) [inline plugins/kubernetes/src/api/types.ts:31]: Let's make this `getObjectsByEntity` instead and pass in the entire entity, instead of breaking it up client-side. That'll provide a lot more flexibility for doing further extensions like this in the future.

Comment (moustafab) [reply plugins/kubernetes/src/api/types.ts]: @Rugvip got back around to this.

File: plugins/kubernetes-backend/src/service/KubernetesFetcher.ts
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesFetcher.ts:112]: when you reach a number of params like this, it's often clearer for callers to have a single object with named keys to pass in instead of separate unnamed arguments

File: plugins/kubernetes-backend/src/service/KubernetesFetcher.ts
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesFetcher.ts:120]: can be condensed to `labelSelector || 'backstage.io/...'` because the empty string is falsy

File: plugins/kubernetes-backend/src/service/KubernetesFetcher.ts
Comment (freben) [inline plugins/kubernetes-backend/src/service/KubernetesFetcher.ts:216]: yeah for real, this is the kind of place where it's super helpful for a reader if this were instead an object - but this is not our code right

File: plugins/kubernetes-backend/src/service/getKubernetesObjectsByServiceIdHandler.test.ts
Comment (freben) [inline plugins/kubernetes-backend/src/service/getKubernetesObjectsByServiceIdHandler.test.ts:28]: redundant double type declaration

File: plugins/kubernetes-backend/src/service/getKubernetesObjectsForServiceHandler.ts
Comment (freben) [inline plugins/kubernetes-backend/src/service/getKubernetesObjectsForServiceHandler.ts:50]: you can also do

```ts
const matchLabels = entity?.spec?.kubernetes?.selector?.matchLabels;
if (matchLabels) {...
```

(but it's pretty defensive to assume that the caller might pass in anything except a `ComponentEntityV1alpha1` :))

File: plugins/kubernetes/README.md
Comment (freben) [inline plugins/kubernetes/README.md:17]: two

File: plugins/kubernetes/README.md
Comment (freben) [inline plugins/kubernetes/README.md:20]: you mean annotation here right?

File: plugins/kubernetes/README.md
Comment (freben) [inline plugins/kubernetes/README.md:18]: in that case it should probably be at the top

**Review by freben — COMMENTED** [review]
Buncha' little things, nothing big

File: plugins/kubernetes/README.md
Comment (moustafab) [reply plugins/kubernetes/README.md:18]: reorganized accordingly

**Comment by moustafab** [comment]
@freben, @Rugvip, @mclarke47 I think it should be good to go, let me know if you'd like me to squash I'll be able to take care of any changes today!

File: packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts
Comment (moustafab) [reply packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts:37]: I tried to do this, but wasn't able with an anonymous type. Should I use the k8s client? Wasn't sure we'd want to introduce that dependency.

Comment (freben) [reply packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts:37]: Yeah the syntax for making generic key/values is a bit funky it seems... Here's an example using `lazy`

https://runkit.com/5908cecfe97ebf0012f2e3c9/59ce587717c5e5001241adac

Essentially it delays the judgment of what to use for checking the values until after the keys are known, so you have the possibility to return `yup.string()` for all of them.

Anyway... Not a big deal really, we can tweak later if it's messy to get it right.

Comment (moustafab) [reply packages/catalog-model/src/kinds/ComponentEntityV1alpha1.ts:37]: Yea this was messier than the value it probably added. Probably something someone who knows typescript better can help with.

**Comment by freben** [comment]
Hey, and thanks for your patience. This is looking good.

My only remaining worry now is, should this be in `spec`. It's not really a property of the component itself, it's more like something that would fit under the annotations concept - but annotations can't be structured data, only strings. So, before merging, let me invoke @mclarke47 @Rugvip @benjdlambert, asking:

Shall we leave this in the spec for now? And potentially reconsider a bit later, to unblock this PR?

Could it be moved verbatim to `metadata`?

Could it be an annotation `backstage.io/kubernetes-match-labels: name=dice-roller,version=canary`? I don't think reserving the comma and equals sign characters would be a problem, and if we in the future relax the data model to support hierarchical data, this string form could be one of two ways of expressing the same thing.

Sorry @moustafab, just a little bit longer, just wanna make sure we considered placement.

**Comment by Rugvip** [comment]
+1 for moving this to an annotation, `backstage.io/kubernetes-match-labels` for now since we can't have structured data in annotations. But eventually it could be moved over to something like `backstage.io/kubernetes-match` if we start allowing structured data in annotations.

**Comment by freben** [comment]
Hey @moustafab,

Thanks again for your work, and apologies again that I bring up this kind of feedback this late. We had a talk over things and agree that yeah this should be a string annotation instead, since it matches the future plans of how this kind of auxiliary data shall be expressed.

Now I don't want to force that upon you, the contents of this PR are fine as they are in all other respects. So I thought I'd leave the choice up to you:

- Would you like to make that change to using an annotation on the form `backstage.io/kubernetes-match-labels: name=dice-roller,version=canary` as part of this PR? I'll try my best to be attentive to reviewing it.
- Or shall we merge this and raise an issue about the change to an annotation, that you'll of course be free to pick up if you feel like it?

**Comment by benjdlambert** [comment]
I wonder if it makes sense to take inspiration from how k8s do structured data in annotations and just json stringify it instead? Then it makes the migration path easier when moving to structured annotations.
https://github.com/backstage/backstage/blob/2fbca83156bf5fa509a347cd5804e41d177b329a/plugins/kubernetes/src/components/Services/__fixtures__/services.json#L5

Maybe it's better than creating a new parser for this field?

**Comment by freben** [comment]
@benjdlambert at this point it's hand written, so that has to be taken into account - it's just too much hassle to type it out (or generate it) by hand like that, including escaping etc

**Comment by freben** [comment]
Re migration, i think it'll end up being of type `string | KubernetesOptions` or whatever in the end, and validated / parsed accordingly.

**Comment by benjdlambert** [comment]
> @benjdlambert at this point it's hand written, so that has to be taken into account - it's just too much hassle to type it out (or generate it) by hand like that, including escaping etc

good point 👍

**Comment by moustafab** [comment]
@freben honestly I think you're probably best off using the annotation with a label query string (similar to how the actual k8s api accepts, and kubectl accepts) and just calling it kubernetes-label-selector. The code would be similar to what I started with and would allow you to include and supply more complex queries that use other verbs than just matchLabels =, like not in or whatever k8s supports. There isn't much value added by making it too complex and it's easier this way if the user knows k8s.

I don't mind taking this on, but just confirm 100% before I do it. I don't want to merge something and start using and then remove it right after.

**Comment by freben** [comment]
That sounds great to me. Good point, no use reinventing anything since that syntax is already human friendly to type out. @mclarke47 final say?

**Comment by mclarke47** [comment]
Sounds good to me (I think this is what I suggested originally before we decided to go with selectors)!

**Comment by freben** [comment]
Cool. @moustafab feel free to continue as works best for you.

**Comment by moustafab** [comment]
Just to confirm plan to pick this back up this weekend and wrap it up.

**Review by benjdlambert — APPROVED** [review]
Nice work! Thanks!

**Comment by mclarke47** [comment]
Great work @moustafab! 🎉

**Comment by moustafab** [comment]
Uh oops, I was actually going to push up the commits to refactor the approach in line with discussion with @freben above... Should I open a new one? I got distracted and couldn't sort out a typescript problem, and so wasn't able to commit the changes 😬 

@mclarke47 @benjdlambert

**Comment by freben** [comment]
@moustafab Sure, just make a new one on top of the new origin master to follow up :)

**Pull request merged**
